### PR TITLE
Add macOS support

### PR DESCRIPTION
**Pull request by amgleitman**
# Summary

This adds macOS support to react-native-svg.

A good amount of code for react-native-svg iOS, such as constructing paths, is already platform-agnostic, but there were a few big things that needed to be resolved, mostly related to type differences between macOS and iOS (for example, importing `<UIKIt/UIKit.h>` and differences between `NSView`/`UIView`, `NSColor`/`UIColor`, etc.), as well as a couple of iOS concepts that don't appear to have direct macOS equivalent (like `-[UIView tintColor]`, which we use to determine what `"currentColor"` is).

Most of the heavy lifting in resolving platform differences lies in RNSVGUIKit.h, which is a similar strategy to what react-native-macos does with RCTUIKit.h (found [here](https://github.com/microsoft/react-native-macos/issues/242)).

## Test Plan

To verify that nothing broke, I created a test app and confirmed that the example SVGs shown in this repo's README render exactly the same.

### Screenshot Comparison

| **macOS** | **iOS** |
| --- | --- |
| Screen Shot 2020-11-30 at 2 19 38 PM | Screen Shot 2020-11-30 at 2 32 06 PM |
| Screen Shot 2020-11-30 at 2 28 50 PM | Screen Shot 2020-11-30 at 2 32 13 PM |
| Screen Shot 2020-11-30 at 2 29 42 PM | Screen Shot 2020-11-30 at 2 32 58 PM |
| Screen Shot 2020-11-30 at 2 30 33 PM | Screen Shot 2020-11-30 at 2 33 04 PM |
| Screen Shot 2020-11-30 at 2 30 41 PM | Screen Shot 2020-11-30 at 2 33 09 PM |

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    N/A     |
| Android |    N/A     |
| macOS   |    ✅     |

## Checklist

- [ ] I have tested this on a device and a simulator
- [ ] I added documentation in `README.md`
- [ ] I updated the typed files (typescript)
- [ ] I added a test for the API in the `__tests__` folder

### CODE REVIEW

File: apple/RNSVGUIKit.h
Comment (HeyImChris) [inline apple/RNSVGUIKit.h:15]: This import can be shared and then we can drop the UIKit/UIKit.h import?

File: apple/RNSVGUIKit.h
Comment (HeyImChris) [inline apple/RNSVGUIKit.h:11]: What's the distinction between RNSVGView and RNSVGPlatformView?

File: apple/RNSVGUIKit.h
Comment (HeyImChris) [inline apple/RNSVGUIKit.h:28]: I'm bad at doing this myself but we should try and pair TODO's with logged issues and then add the issue bug number as a comment here

File: .eslintignore
Comment (HeyImChris) [inline .eslintignore:7]: Could you add a screenshot of the macOS SVG images to the PR :)?

File: apple/RNSVGUIKit.h
Comment (amgleitman) [reply apple/RNSVGUIKit.h:15]: Unfortunately, we can't do this until RCTUIKit.h becomes available in react-native.

File: apple/RNSVGUIKit.h
Comment (amgleitman) [reply apple/RNSVGUIKit.h:11]: RNSVGView is meant to be an extension of RCTUIView and includes all the bells and whistles that we would expect from a React Native view. RNSVGPlatformView is an alias for a platform-specific view, which is useful for functions that are already built into AppKit/UIKit, such as traveling through the view hierarchy. (For example, look at `-[RNSVGView tintColor]` in RNSVGUIKit.macos.m.)

File: apple/RNSVGUIKit.h
Comment (amgleitman) [reply apple/RNSVGUIKit.h:28]: Done.

File: .eslintignore
Comment (amgleitman) [reply .eslintignore:7]: Done.

File: RNSVG.podspec
Comment (HeyImChris) [inline RNSVG.podspec:6]: Why are all these lines marked as a diff?

File: apple/RNSVGUIKit.h
Comment (HeyImChris) [reply apple/RNSVGUIKit.h:11]: But since RCTUIView inherits from UIView and NSView can we just use that in place of RNSVGPlatformView?

File: apple/Elements/RNSVGSvgView.m
Comment (HeyImChris) [inline apple/Elements/RNSVGSvgView.m:29]: Do we need this TODO if it's working on Mac?

File: apple/RNSVGUIKit.macos.m
Comment (HeyImChris) [inline apple/RNSVGUIKit.macos.m:1]: Instead of the entire file being in an if def can we just only build it for the macOS target?

File: apple/Text/RNSVGTopAlignedLabel.ios.m
Comment (HeyImChris) [inline apple/Text/RNSVGTopAlignedLabel.ios.m:1]: Just get rid of the if def and only build into iOS targets?

**Review by HeyImChris — CHANGES_REQUESTED** [review]

File: RNSVG.podspec
Comment (amgleitman) [reply RNSVG.podspec:6]: Because we're adding an additional space to account for `s.ios.exclude_files` and `s.osx.exclude_files`.

File: apple/RNSVGUIKit.h
Comment (amgleitman) [reply apple/RNSVGUIKit.h:11]: RCTUIView is defined in RCTUIKit.h, which isn't available on plain react-native yet, so we can't use RCTUIView.

File: apple/RNSVGUIKit.macos.m
Comment (amgleitman) [reply apple/RNSVGUIKit.macos.m:1]: We already exclude *.macos.m files on iOS as per the podspec file. I just threw the ifdef in as a "just in case" sort of thing, but yeah, we technically don't need it.

File: apple/Text/RNSVGTopAlignedLabel.ios.m
Comment (amgleitman) [reply apple/Text/RNSVGTopAlignedLabel.ios.m:1]: We already exclude *.ios.m files on macOS as per the podspec file. I just threw the ifdef in as a "just in case" sort of thing, but yeah, we technically don't need it.

File: apple/Elements/RNSVGSvgView.m
Comment (amgleitman) [reply apple/Elements/RNSVGSvgView.m]: Discussed offline. Since macOS appears to work fine, I'm simply removing this TODO.

File: apple/RNSVGUIKit.macos.m
Comment (HeyImChris) [reply apple/RNSVGUIKit.macos.m]: Can we delete it if we don't need it? If the code has no function then let's keep things clean and not have it

Comment (HeyImChris) [reply apple/RNSVGUIKit.macos.m]: Same request for the iOS file

**Review by HeyImChris — APPROVED** [review]

File: apple/RNSVGUIKit.macos.m
Comment (amgleitman) [reply apple/RNSVGUIKit.macos.m]: Already done with this commit:
https://github.com/react-native-svg/react-native-svg/pull/1494/commits/a35679a44db0982415b2390ed1e6bf8ab0541bc1

**Review by Saadnajmi — APPROVED** [review]

**Comment by jacobp100** [comment]
@amgleitman Thank you so much for working on this! Looks like a huge undertaking. It's been one of the things holding me back from using rn-macos and this will help a lot

**Comment by amgleitman** [comment]
@msand, can you take a look at this?

**Comment by jacobp100** [comment]
@amgleitman Do you have a published version of https://github.com/microsoft/react-native-svg-macos?

**Comment by amgleitman** [comment]
> @amgleitman Do you have a published version of https://github.com/microsoft/react-native-svg-macos?

@jacobp100 Not yet, but we're working on it.

**Comment by jacobp100** [comment]
Ok. Thanks for your work on this! It'll be really helpful

**Comment by msand** [comment]
This is really impressive work! I hope to have some time and energy to dedicate to open source soon, becoming a father this spring, and our startup is growing quickly atm...

**Comment by HeyImChris** [comment]
> This is really impressive work! I hope to have some time and energy to dedicate to open source soon, becoming a father this spring, and our startup is growing quickly atm...

To give us some flexibility with contributing to this repo, do you think it'd be possible to give @amgleitman write permissions so he can merge in the Mac support change? Either way, congrats on the startup and soon-to-be fathering!

**Comment by vincentjames501** [comment]
This is awesome! @msand do you have any thoughts to @HeyImChris 's question?

**Comment by msand** [comment]
I just added @amgleitman to the repository with write rights. Think I'll announce in a separate issue that this repo is looking for more maintainers and add a link to the readme. Thanks for any help in this!

**Comment by amgleitman** [comment]
Thanks @msand! Can I assume that I have your blessing to merge this PR and help maintain this repo in the future?

**Comment by vomchik** [comment]
@amgleitman do you have the source code of the SVG content(all you show on screenshots)? I want to try to write tests for this library.

**Comment by amgleitman** [comment]
> @amgleitman do you have the source code of the SVG content(all you show on screenshots)? I want to try to write tests for this library.

I took the examples from the README file.

**Comment by vomchik** [comment]
> > @amgleitman do you have the source code of the SVG content(all you show on screenshots)? I want to try to write tests for this library.
> 
> I took the examples from the README file.

i see :) thanks

**Pull request merged**
